src/KubernetesClient/Watch.cs

```diff
@@ -0,0 +1,482 @@
+            if (!gotType) throw new JsonSerializationException("The stream does not appear to contain watch events.");
+            return e;
+        }
+
+        #region CancelableStreamReader
+        // HACK: annoyingly, the TextReader interface doesn't support cancellation of async reads outside .NET core. furthermore, Json.NET
+        // doesn't support cancellation even in several versions of .NET core. so we have this hacky TextReader that attempts cancellation
+        sealed class CancelableStreamReader : TextReader
+        {
+            public CancelableStreamReader(Stream stream) => this.stream = stream;
+
+            // the CancellationToken used to cancel reads
+            public CancellationToken CancelToken { get; set; }
+
+            public override int Peek() => EnsureData() ? charBuffer[charRead] : -1;
+            public override int Read() => EnsureData() ? charBuffer[charRead++] : -1;
+            public override int Read(char[] buffer, int index, int count) => Read(buffer, index, count, false);
+
+#if NETCOREAPP2_1
+            public override ValueTask<int> ReadAsync(Memory<char> buffer, CancellationToken _) => ReadAsync(buffer, false);
+
+            public override Task<int> ReadAsync(char[] buffer, int index, int count) =>
+                ReadAsync(buffer.AsMemory(index, count), false).AsTask();
+
+            public override ValueTask<int> ReadBlockAsync(Memory<char> buffer, CancellationToken _) => ReadAsync(buffer, true);
+
+            public override Task<int> ReadBlockAsync(char[] buffer, int index, int count) =>
+                ReadAsync(buffer.AsMemory(index, count), true).AsTask();
+#else
+            public override Task<int> ReadAsync(char[] buffer, int index, int count) => ReadAsync(buffer, index, count, false);
+            public override Task<int> ReadBlockAsync(char[] buffer, int index, int count) => ReadAsync(buffer, index, count, true);
+#endif
+
+            public override string ReadToEnd()
+            {
+                var sb = new System.Text.StringBuilder(4096);
+                while(EnsureData())
+                {
+                    sb.Append(charBuffer, charRead, charWrite-charRead);
+                    charRead = charWrite;
+                }
+                return sb.ToString();
+            }
+
+            public async override Task<string> ReadToEndAsync()
+            {
+                var sb = new System.Text.StringBuilder(4096);
+                while(await EnsureDataAsync().ConfigureAwait(false))
+                {
+                    sb.Append(charBuffer, charRead, charWrite-charRead);
+                    charRead = charWrite;
+                }
+                return sb.ToString();
+            }
+
+            protected override void Dispose(bool disposing)
+            {
+                base.Dispose(disposing);
+                stream?.Dispose();
+            }
+
+            bool EnsureData()
+            {
+                if (charRead == charWrite)
```

Comment:
Consider delegating to `EnsureDataAsync` instead:
```csharp
if (charRead == charWrite)
{
  try { return EnsureDataAsync().GetAwaiter().GetResult(); }
  catch (OperationCanceledException) { return false; } // non-async callers aren't expecting OperationCanceledException
}
```

Reply:
I considered it but liked having the fast path not create a task. But your suggestion of only doing it if we would do a read anyway is better. I'll do that.